### PR TITLE
Dow 27 cache service

### PR DESCRIPTION
**Pull request by mnojokes**
Implemented initial CacheService logic (waiting for db repository to finalize). Updated PeriodicHostedService to take period time span from appsettings.json.

### CODE REVIEW

File: DownloadMoreRAM/src/API/Controllers/CacheController.cs [hunk 1]
```diff
@@ -1,35 +1,44 @@
...
+        await _cacheService.Add(key, data);
+        return NoContent();
     }
 
     [HttpPut("{key}")]
     public async Task<IActionResult> Append([FromHeader(Name = "X-Api-Key")] string userApiKey, string key, [FromBody] CacheData data)
     {
-        throw new NotImplementedException();
+        await _cacheService.Add(key, data);
+        return NoContent();
```
Comment (rusvaplauke) [inline DownloadMoreRAM/src/API/Controllers/CacheController.cs:29]: Same comment as before: a good idea might be to return Ok + the updated object.

File: DownloadMoreRAM/src/API/Controllers/CacheController.cs [hunk 2]
```diff
@@ -1,35 +1,44 @@
...
     [HttpPost("{key}")]
     public async Task<IActionResult> Create([FromHeader(Name = "X-Api-Key")] string userApiKey, string key, [FromBody] CacheData data) 
     {
-        throw new NotImplementedException();
+        await _cacheService.Add(key, data);
+        return NoContent();
     }
 
     [HttpPut("{key}")]
     public async Task<IActionResult> Append([FromHeader(Name = "X-Api-Key")] string userApiKey, string key, [FromBody] CacheData data)
```
Comment (rusvaplauke) [inline DownloadMoreRAM/src/API/Controllers/CacheController.cs:26]: If we use the CacheData dto for this, we also demand user for the ExpiryDate. Is that in line with the requirements - i.e., shouldn't we decide on this period ourselves (based on their original input OR based on the max period)? I'm open for dicussions!

File: DownloadMoreRAM/src/API/Controllers/CacheController.cs [hunk 3]
```diff
@@ -1,35 +1,44 @@
...
 {
+    private readonly CacheService _cacheService;
+
+    public CacheController(CacheService cacheService)
+    {
+        _cacheService = cacheService;
+    }
+
     [HttpPost("{key}")]
     public async Task<IActionResult> Create([FromHeader(Name = "X-Api-Key")] string userApiKey, string key, [FromBody] CacheData data)
```
Comment (rusvaplauke) [inline DownloadMoreRAM/src/API/Controllers/CacheController.cs:19]: CacheData carries the expiration DATE, whereas my understanding from the requirements was that we ask the user to enter the expiration PERIOD (i.e., "I want this to be valid for 10 minutes" -->  "created values should have expiration period that should be received during 'create' call. If expiration period is not defined - default value should be used from config"). 
We can discuss and interpret in another way, of course.

File: DownloadMoreRAM/src/API/Controllers/CacheController.cs [hunk 4]
```diff
@@ -1,35 +1,44 @@
 ﻿using Microsoft.AspNetCore.Mvc;
 using Domain.DTO;
+using Application.Services;
 
 namespace API.Controllers;
 
 [ApiController]
 [Route("[controller]")]
```
Comment (rusvaplauke) [inline DownloadMoreRAM/src/API/Controllers/CacheController.cs:8]: Naming: is "cache" the resource we're dealing with, or is it individual objects that are in the cache (i.e., "I want to delete cache" vs "I want to delete smth smth from the cache")

File: DownloadMoreRAM/src/Application/Services/CacheService.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    // private readonly ICacheRepository _cacheRepository;
+
+    public CacheService(IConfiguration configuration)
+    {
+        double expiryDuration;
+        if (!double.TryParse(configuration["MaxExpiryDuration"], out expiryDuration))
+        {
+            throw new ArgumentNullException("MaxExpiryDuration");
+        }
+        _maxExpiryDuration = TimeSpan.FromSeconds(expiryDuration);
```
Comment (rusvaplauke) [inline DownloadMoreRAM/src/Application/Services/CacheService.cs:20]: shouldn't this be converted to minutes, etc? Or will we delete a record after max 20 secs?

File: DownloadMoreRAM/src/Application/Services/CacheService.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        // TODO: add repository dependency
+        _lock.WaitOne();
+        // TODO: populate _cache with data from DB
+        _lock.ReleaseMutex();
+    }
+
+    public async Task Add(string key, CacheData data)
+    {
+        _lock.WaitOne();
+        // TODO: convert CacheData to CacheEntity (check with _maxExpiryDuration) and send to repo
```
Comment (rusvaplauke) [inline DownloadMoreRAM/src/Application/Services/CacheService.cs]: missing check if entry with the key already exists

File: DownloadMoreRAM/src/Application/Services/CacheService.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        _lock.ReleaseMutex();
+        throw new NotImplementedException();
+    }
+
+    public async Task Delete(string key)
+    {
+        _lock.WaitOne();
+        await Delete_NonBlocking(key);
```
Comment (rusvaplauke) [inline DownloadMoreRAM/src/Application/Services/CacheService.cs]: missing check if exists

**Review by rusvaplauke — CHANGES_REQUESTED** [review]
added some comments, will continue in the office

**Review by ErikasSilobritas — APPROVED** [review]

**Pull request merged**